frontend/src/App.js

```diff
@@ -40,34 +68,76 @@ FROM ( SELECT MONTH(committer_when) as month,
 
     this.showCode = this.showCode.bind(this);
     this.showUAST = this.showUAST.bind(this);
+  }
+
+  setState(partialState, callback) {
+    super.setState(partialState, callback);
 
-    this.uniqueKey = 0;
+    if (typeof partialState.results !== 'undefined') {
+      persist('results', partialState.results);
+    }
+    if (typeof partialState.history !== 'undefined') {
+      persist('history', partialState.history);
+    }
   }
 
   handleTextChange(text) {
     this.setState({ sql: text });
   }
 
   setResult(key, result) {
-    if (!this.state.results.has(key)) {
+    const { results, history } = this.state;
+    const historyIdx = history.findIndex(i => i.key === key);
+
+    if (historyIdx >= 0) {
+      const status =
+        typeof result.response !== 'undefined' ? STATUS_SUCCESS : STATUS_ERROR;
+      const newHistory = [
+        ...history.slice(0, historyIdx),
+        {
+          ...history[historyIdx],
+          status,
+          errorMsg: result.errorMsg
+        },
+        ...history.slice(historyIdx + 1)
+      ];
+      this.setState({ history: newHistory });
+    }
+
+    if (!results[key]) {
       // Tab was removed, ignore results
       return;
     }
 
-    const newResults = new Map(this.state.results);
-    newResults.set(key, result);
+    const newResults = {
+      ...results,
+      [key]: result
+    };
 
     this.setState({ results: newResults });
   }
 
   handleSubmit() {
-    const { sql } = this.state;
-    const key = ++this.uniqueKey;
+    const { sql, history } = this.state;
+    const key = nanoid();
 
-    const loadingResults = new Map(this.state.results);
-    loadingResults.set(key, { sql, loading: true });
+    const loadingResults = {
+      ...this.state.results,
+      [key]: { sql, loading: true }
+    };
 
-    this.setState({ results: loadingResults });
+    this.setState({
+      results: loadingResults,
+      history: [
+        {
+          key,
+          sql,
+          datetime: new Date(),
+          status: STATUS_LOADING
```

Comment:
Could you please help me understand, if a tab title and all the other fields from https://github.com/src-d/gitbase-playground/issues/71 get persisted in the history entities?

Reply:
a current title is sql and as you can see I persist it. We will have a separate field for the title in #72